File: mypay.sources/mypay4-be/src/main/java/it/regioneveneto/mygov/payment/mypay4/service/fesp/EnteService.java
/**
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package it.regioneveneto.mygov.payment.mypay4.service.fesp;

import it.regioneveneto.mygov.payment.mypay4.dao.fesp.EnteDao;
import it.regioneveneto.mygov.payment.mypay4.model.fesp.Ente;
import it.regioneveneto.mygov.payment.mypay4.service.common.CacheService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service("fespEnteService")
public class EnteService {

  @Autowired
  private EnteDao enteDao;

  @Cacheable(value=CacheService.CACHE_NAME_FESP_ENTE, key="{'mygovEnteId',#mygovEnteId}", unless="#result==null")
  @Transactional(transactionManager = "tmFesp", propagation = Propagation.SUPPORTS)
  public Ente getEnteById(Long mygovEnteId) {
    return enteDao.getEnteById(mygovEnteId);
  }

  @Cacheable(value=CacheService.CACHE_NAME_FESP_ENTE, key="{'codIpa',#codIpa}", unless="#result==null")
  @Transactional(transactionManager = "tmFesp", propagation = Propagation.SUPPORTS)
  public Ente getEnteByCodIpa(String codIpa) {
    return enteDao.getEnteByCodIpa(codIpa);
  }

  @Cacheable(value=CacheService.CACHE_NAME_FESP_ENTE, key="{'codFiscale',#codFiscale}", unless="#result==null")
  @Transactional(transactionManager = "tmFesp", propagation = Propagation.SUPPORTS)
  public Ente getEnteByCodFiscale(String codFiscale) {
   return enteDao.getEnteByCodFiscale(codFiscale);
  }

  @Transactional(transactionManager = "tmFesp", propagation = Propagation.SUPPORTS)
  public List<Ente> getAllEnti() {
    return enteDao.getAllEnti();
  }

}

File: mypay.sources/mypay4-fe/projects/mypay4-fe-operatore/src/app/services/ente.service.ts
/*
 *     MyPay - Payment portal of Regione Veneto.
 *     Copyright (C) 2022  Regione Veneto
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
import {
    ApiInvokerService, BaseUrlService, Ente, LocalCacheService, TipoDovuto, UserService
} from 'projects/mypay4-fe-common/src/public-api';
import { BehaviorSubject, Observable, of, Subject, Subscription } from 'rxjs';
import { filter, map } from 'rxjs/operators';

import { Injectable, OnDestroy } from '@angular/core';
import { Router, RoutesRecognized } from '@angular/router';

@Injectable({
  providedIn: 'root'
})
export class EnteService implements OnDestroy{

  private currentEnte: Ente;
  private currentEnteSubject: Subject<Ente> = new BehaviorSubject(null);
  private userChangeSub: Subscription;
  private needEnte: boolean;
  private needEnteSubject: Subject<boolean> = new BehaviorSubject(false);
  private routeChangeSub: Subscription;

  private entiForLogo = new Map<number, Ente>();

  constructor(
    private apiInvokerService: ApiInvokerService,
    private userService: UserService,
    private baseUrlService: BaseUrlService,
    private router: Router,
    private localCacheService: LocalCacheService,
  ) {
    //reset current ente if user changes
    this.userChangeSub = this.userService.getLoggedUserObs().subscribe(user => {
      this.setCurrentEnte(null);
    });
    this.routeChangeSub = this.router.events
      .pipe(filter(event => event instanceof RoutesRecognized))
      .subscribe( (event:RoutesRecognized) => {
        const data = event.state.root.firstChild.data;
        let needEnte = true;
        if(data.needEnte != null)
          needEnte = data.needEnte;
        else if(data.menuItem?.length>0)
          needEnte = data.menuItem[0].needEnte; //if more menuItem, just chek the first (they are supposed to be all the same value)
        this.setNeedEnte(needEnte);
      });
  }

  ngOnDestroy():void {
    this.userChangeSub?.unsubscribe();
    this.routeChangeSub?.unsubscribe();
  }

  getCurrentEnte():Ente {
    return this.currentEnte;
  }

  setCurrentEnte(ente:Ente) {
    const changed = this.currentEnte !== ente;
    this.currentEnte = ente;
    if(changed || this.currentEnte === null){
      this.currentEnteSubject.next(this.currentEnte);
    }
  }

  getCurrentEnteObs():Observable<Ente> {
    return this.currentEnteSubject;
  }

  isNeedEnte(): boolean {
    return this.needEnte;
  }

  setNeedEnte(needEnte: boolean) {
    const changed = this.needEnte !== needEnte;
    this.needEnte = needEnte;
    if(changed){
      this.needEnteSubject.next(this.needEnte);
    }
  }

  getNeedEnteObs():Observable<boolean> {
    return this.needEnteSubject;
  }

  getAllEnti(): Observable<Ente[]> {
    return this.localCacheService.manageThumbLogoEntiCache(
      this.apiInvokerService.get<Ente[]>(this.baseUrlService.getPubOperatoreUrl() + 'enti?logoMode=hash'),
      (apiInvokerServiceRef: ApiInvokerService) => apiInvokerServiceRef.get<Ente[]>(this.baseUrlService.getPubOperatoreUrl() + 'enti?logoMode=thumb')
    );
  };

  getEntiByOperatoreUserId(): Observable<Ente[]> {
    return this.localCacheService.manageThumbLogoEntiCache(
      this.apiInvokerService.get<Ente[]>(this.baseUrlService.getOperatoreUrl() + 'enti?logoMode=hash', {skipHandleError: true}),
      (apiInvokerServiceRef: ApiInvokerService) => apiInvokerServiceRef.get<Ente[]>(this.baseUrlService.getOperatoreUrl() + 'enti?logoMode=thumb', {skipHandleError: true})
    );
  }


  getListTipoDovutoByEnte(ente: Ente): Observable<TipoDovuto[]> {
    return this.apiInvokerService.get<TipoDovuto[]>(this.baseUrlService.getBaseUrlApi() + 'enti/' + ente.mygovEnteId + '/tipiDovuto');
  }

  getListTipoDovutoByEnteAsOperatore(ente: Ente): Observable<TipoDovuto[]> {
    return this.apiInvokerService.get<TipoDovuto[]>(this.baseUrlService.getOperatoreUrl() + 'enti/' + ente.mygovEnteId + '/tipiDovutoOperatore');
  }

  getEnte(idEnte: number, thumbnail:boolean = true): Observable<Ente> {
    return this.entiForLogo.has(idEnte) ?
      of(this.entiForLogo.get(idEnte)) :
      this.apiInvokerService.get<Ente>(this.baseUrlService.getPubCittadinoUrl() + 'enti/' + idEnte)
        .pipe(
          map(ente => {
            this.entiForLogo.set(idEnte, ente);
            return ente;
          })
        );
  }

}
